src/utils.ts

```diff
@@ -26,3 +32,91 @@ export function toWei(
 export function fromWei(amount: BigNumber, decimals: number): BigNumber {
   return amount.dividedBy(TEN.pow(decimals));
 }
+
+export type SummaryEntry = {
+  tokenAddress: string;
+  amount: BigNumber;
+  decimals: number;
+};
+
+export const transfersToSummary = (transfers: Payment[]) => {
+  return transfers.reduce((previousValue, currentValue): Map<
+    string,
+    SummaryEntry
+  > => {
+    let tokenSummary = previousValue.get(currentValue.tokenAddress);
+    if (typeof tokenSummary === "undefined") {
+      tokenSummary = {
+        tokenAddress: currentValue.tokenAddress,
+        amount: new BigNumber(0),
+        decimals: currentValue.decimals,
+      };
+      previousValue.set(currentValue.tokenAddress, tokenSummary);
+    }
+    tokenSummary.amount = tokenSummary.amount.plus(currentValue.amount);
+
+    return previousValue;
+  }, new Map<string, SummaryEntry>());
+};
+
+export type InsufficientBalanceInfo = {
+  token: string;
+  transferAmount: string;
+};
+
+export const checkAllBalances = async (
+  summary: Map<string, SummaryEntry>,
+  web3Provider: ethers.providers.Web3Provider,
+  safe: SafeInfo,
+  tokenList: TokenMap
+): Promise<InsufficientBalanceInfo[]> => {
+  const insufficientTokens: InsufficientBalanceInfo[] = [];
+  for (const summaryEntry of summary.values()) {
+    const tokenAddress = summaryEntry.tokenAddress;
+    const tokenAmount = summaryEntry.amount;
+    if (tokenAddress === null) {
+      // Check ETH Balance
+      const tokenBalance = await web3Provider.getBalance(
+        safe.safeAddress,
+        "latest"
+      );
+      if (!isSufficientBalance(tokenBalance, tokenAmount, 18)) {
+        insufficientTokens.push({
+          token: "ETH",
+          transferAmount: tokenAmount.toFixed(),
+        });
+      }
+    } else {
+      const erc20Contract = erc20Instance(
+        utils.getAddress(tokenAddress),
+        web3Provider
+      );
+      console.log(erc20Contract);
+      const tokenBalance = await erc20Contract.balanceOf(safe.safeAddress);
+      const tokenInfo = tokenList.get(tokenAddress);
+      if (
+        !isSufficientBalance(
+          tokenBalance,
+          tokenAmount,
+          tokenInfo?.decimals || summaryEntry.decimals
```

Comment:
Is there a reason a particular scenario you have in mind here for when tokenInfo won't have the decimals and we need to use the summaryEntry? Is it just for ETH when `tokenAddress === null`?

Reply:
It is for any token address not present in the list. For those tokenInfo will be undefined and thus tokenInfo?.decimals will also be undefined.